src/main/java/moze_intel/projecte/integration/jei/world_transmute/WorldTransmuteEntry.java

```diff
@@ -0,0 +1,155 @@
+package moze_intel.projecte.integration.jei.world_transmute;
+
+import mezz.jei.api.ingredients.IIngredients;
+import mezz.jei.api.recipe.IRecipeWrapper;
+import moze_intel.projecte.utils.WorldTransmutations;
+import net.minecraft.block.Block;
+import net.minecraft.block.state.IBlockState;
+import net.minecraft.client.resources.I18n;
+import net.minecraft.init.Blocks;
+import net.minecraft.item.ItemStack;
+import net.minecraftforge.fluids.FluidRegistry;
+import net.minecraftforge.fluids.FluidStack;
+
+import javax.annotation.Nonnull;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+public class WorldTransmuteEntry implements IRecipeWrapper
+{
+	private ItemStack inputItem = ItemStack.EMPTY;
+	private ItemStack leftOutputItem = ItemStack.EMPTY;
+	private ItemStack rightOutputItem = ItemStack.EMPTY;
+	private FluidStack inputFluid;
+	private FluidStack leftOutputFluid;
+	private FluidStack rightOutputFluid;
+
+	public WorldTransmuteEntry(WorldTransmutations.Entry transmutationEntry)
+	{
+		Block inputBlock = transmutationEntry.input.getBlock();
+		IBlockState leftOutput = transmutationEntry.outputs.getLeft();
+		IBlockState rightOutput = transmutationEntry.outputs.getRight();
+
+		inputFluid = fluidFromBlock(inputBlock);
+		if (inputFluid == null)
+		{
+			inputItem = itemFromBlock(inputBlock, transmutationEntry.input);
+		}
+		if (leftOutput != null)
+		{
+			leftOutputFluid = fluidFromBlock(leftOutput.getBlock());
+			if (leftOutputFluid == null)
+			{
+				leftOutputItem = itemFromBlock(leftOutput.getBlock(), leftOutput);
+			}
+		}
+		if (rightOutput != null)
+		{
+			rightOutputFluid = fluidFromBlock(rightOutput.getBlock());
+			if (rightOutputFluid == null)
+			{
+				rightOutputItem = itemFromBlock(rightOutput.getBlock(), rightOutput);
+			}
+		}
+	}
+
+	private FluidStack fluidFromBlock(Block block)
+	{
+        if (block == Blocks.WATER)
+        {
+            return new FluidStack(FluidRegistry.WATER, 1000);
+        }
+        else if (block == Blocks.LAVA)
+        {
+            return new FluidStack(FluidRegistry.LAVA, 1000);
+        }
+		return null;
+	}
+
+	private ItemStack itemFromBlock(Block block, IBlockState state)
```

Comment:
why not just use statetostack?

Reply:
Because it does not work quite properly for this case. It gives "invalid" item stacks for things like rotated pumpkins, or non decayable leaves. Attempting to use pickBlock makes it get the proper type, though I did change the fallback to use statetostack, as I did not realize that was a method when I was writing this. (Discovered its existence while rewriting the mercurial eye)